keyboards/handwired/replicazeron/common/oled.c

```diff
@@ -0,0 +1,96 @@
+/* Copyright 2023 9R
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "oled.h"
+
+uint8_t shiftbits =32 ;
+
+//////////// OLED output helpers //////////////
+void draw_mode(controller_state_t controller_state) {
+    //draw oled row showing thumbstick mode
+    oled_write_P(PSTR("Mode: "), false);
+    if (controller_state.wasdShiftMode) {
+        oled_write_ln_P(PSTR("WASD + Shift"), false);
+    } else if (controller_state.wasdMode) {
+        oled_write_ln_P(PSTR("WASD"), false);
+    } else {
+        oled_write_ln_P(PSTR("JoyStick"), false);
+    }
+}
+
+void draw_wasd_key(wasd_state_t wasd_state) {
+    //draw oled row showing active keypresses emulated from thumbstick
+    const char* keys = "wasd";
+    bool keystates [] = { wasd_state.w, wasd_state.a, wasd_state.s, wasd_state.d };
+    // iterate over keystates
+    for (uint8_t i = 0 ; i < ARRAY_SIZE(keystates); ++i) {
+        if (keystates[i]) {
+            char k = keys[i] ;
+            //bitshift char to upper case
+            if (wasd_state.shift) {
+                k &= ~shiftbits;
+            }
+            sprintf(stringbuffer, "%c", k);
```

Comment:
sprintf is expensive for avr.

That said, if k is just the character, you should be able to use `oled_write`, or `oled_write_char`

Reply:
Thank you for the help @drashna . I tested your suggestion and it works as expected. Changes have been aplied in 649af0155f5c90379cd101d5e8bdc47a38c1126d